crates/ty_python_semantic/src/types/newtype.rs

```diff
@@ -0,0 +1,305 @@
+                        db,
+                        self.name(db).clone(),
+                        self.definition(db),
+                        Some(mapped_base),
+                    );
+                }
+            }
+        }
+        // If we get here, there is no base class (because this newtype is cyclic), and we don't
+        // call `f` at all.
+        self
+    }
+
+    pub(crate) fn normalized_impl(self, db: &'db dyn Db, visitor: &NormalizedVisitor<'db>) -> Self {
+        self.map_base_class_type(db, |base_class_type| {
+            // We don't really want to work with `Type` directly (`NewTypeBase` is much more
+            // restricted), but we need to let the `NormalizedVisitor` see the `Type` of the base
+            // class, because this is its only chance to detect an indirect cycle like this one:
+            //
+            // ```py
+            // Foo = NewType("Foo", list["Foo"])
+            // ```
+            //
+            // Note that NewTypeBaseIter does *not* catch this kind of cycle, because it stops
+            // after returning `NewTypeBase::ClassType` for `list`.
+            let mut is_first_encounter = false;
+            let base_type = Type::instance(db, base_class_type);
+            _ = visitor.visit(base_type, || {
+                // If we hit in a cycle, this closure won't run.
+                is_first_encounter = true;
+                base_type
+            });
+            if is_first_encounter {
+                base_class_type.normalized_impl(db, visitor)
+            } else {
+                // Cycle detected. Fall back to `object`.
+                // REVIEWERS: Is this correct?
+                ClassType::object(db)
+            }
+        })
+    }
+
+    pub(crate) fn apply_type_mapping_impl<'a>(
+        self,
+        db: &'db dyn Db,
+        type_mapping: &TypeMapping<'a, 'db>,
+        tcx: TypeContext<'db>,
+        visitor: &ApplyTypeMappingVisitor<'db>,
+    ) -> Self {
+        self.map_base_class_type(db, |base_class_type| {
+            // We don't really want to work with `Type` directly (`NewTypeBase` is much more
+            // restricted), but we need to let the `ApplyTypeMappingVisitor` see the `Type` of the
+            // base class, because this is its only chance to detect an indirect cycle like:
+            //
+            // ```py
+            // Foo = NewType("Foo", list["Foo"])
+            // ```
+            //
+            // Note that NewTypeBaseIter does *not* catch this kind of cycle, because it stops
+            // after returning `NewTypeBase::ClassType` for `list`.
+            let mut is_first_encounter = false;
+            let base_type = Type::instance(db, base_class_type);
+            _ = visitor.visit(base_type, || {
+                // If we hit in a cycle, this closure won't run.
+                is_first_encounter = true;
+                base_type
+            });
+            if is_first_encounter {
```

Comment:
@AlexWaygood One of the downsides of using the restricted `NewTypeBase` enum instead of just making the base a `Type`, is that calling `visitor.visit()` here isn't "natural". We need to synthesize a `Type` just to be able to make this call for its side effects, and then we discard the return value. But if we omit this (and of course I omitted it at first), the last line of the mdtests crashes with a stack overflow. I wonder if that's an argument for making our visitor machinery more flexible somehow?

(Incidentally `normalized_impl` is in the exact same situation, but I don't think I have a test case to provoke a stack overflow for it. Any suggestions?)

Reply:
Ohhhh that does make sense.